### PR TITLE
[FIX] web_editor: page getting crash while creating table

### PR DESCRIPTION
**Pull request by adsa-odoo**
**Current behaviour before commit:**

When text is selected, 
1. applying table command through powerbox throws error.
2. applying table command through powerbox makes toolbar 
appear along with tablepicker.

**Desired behaviour after commit:**

Now when text is selected, 
1. applying table command through powerbox creates new table 
without any traceback.
2. applying table command through powerbox opens tablepicker only.

task-3482230

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 1]
```diff
@@ -2329,7 +2329,7 @@ export function getRangePosition(el, document, options = {}) {
         clonedRange.detach();
     }
 
-    if (!offset || offset.heigh === 0) {
+    if (!offset || offset.height === 0) {
```
Comment (shsa-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2332]: is it related to the fix?

Comment (adsa-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2332]: @shsa-odoo No, but I noticed that it may cause a bug in future and it is fixed in 16.0 and above.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 2]
```diff
@@ -2423,7 +2423,7 @@ export function rgbToHex(rgb = '') {
 
 export function getRangePosition(el, document, options = {}) {
     const selection = document.getSelection();
-    if (!selection.isCollapsed || !selection.rangeCount) return;
+    if (!selection.rangeCount) return;
```
[the remark this answers is not in the document]
Comment (adsa-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2438]: I've done changes. Can you please have a look :)

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 3]
```diff
@@ -352,7 +352,7 @@ export class Powerbox {
     }
 
     _resetPosition() {
-        const position = getRangePosition(this.el, this.options.document, this.options);
+        const position = getRangePosition(this.el, this.options.document, { parentContextRect: this.options.getContextFromParentRect() });
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:355]: Was this for sure the only option we were interested in? This seems like an unrelated and possibly breaking change.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 4]
```diff
@@ -2420,16 +2420,29 @@ export function rgbToHex(rgb = '') {
...
+ * @param {Document} document
+ * @param {Object} [options]
+ * @param {Number} [options.marginRight] right margin to be considered
+ * @param {Number} [options.marginBottom] bottom margin to be considered
+ * @param {Number} [options.marginTop] top margin to be considered
+ * @param {Number} [options.marginLeft] left margin to be considered
+ * @param {DOMRect} [options.parentContextRect] context rect from parent
+ * @returns {Object | undefined}
+ */
+export function getRangePosition(el, document, { marginRight, marginBottom, marginTop, marginLeft, parentContextRect } = {}) {
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2437]: We can't change the API in stable like this, in case someone was using it in their customization. Please only change the docstring without changing the arguments of the function.

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js (hunk 3 above)
Comment (adsa-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:355]: Till now, Yes! Actually it is a fix for a bug. See this [commit](https://github.com/odoo/odoo/pull/104960/commits/8795d335643a9ed7bbac7df046e1abef1878596a). When the function was originally created we needed four margin values as options which we have not used yet (Might be because we didn't find need for it). Right now the same bug can be reproduced in case of table picker, [See](https://www.awesomescreenshot.com/image/46172078?key=fbea4539921d7a8e89729431053378ed). I have tried to cover it by doing the same thing. What do you think would be the best alternate action here?

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js (hunk 4 above)
Comment (adsa-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js]: Thanks for letting me know. I'll keep that in mind :+1: .

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js (hunk 3 above)
Comment (Zynton) [reply addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:355]: What were we passing in `this.options`? Obviously `getContextFromParentRect` was already in it, so is this meant to remove the margin values? I don't understand what this change does.

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 5]
```diff
@@ -2420,10 +2420,23 @@ export function rgbToHex(rgb = '') {
...
+ * position of a range in case of non-collapsed range).
+ *
+ * @param {HTMLElement} el element for which range postion will be calculated
+ * @param {Document} document
+ * @param {Object} [options]
+ * @param {Number} [options.marginRight] right margin to be considered
+ * @param {Number} [options.marginBottom] bottom margin to be considered
+ * @param {Number} [options.marginTop] top margin to be considered
+ * @param {Number} [options.marginLeft] left margin to be considered
+ * @param {DOMRect} [options.parentContextRect] context rect from parent
```
Comment (Zynton) [inline addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2434]: I consider that expecting different keys in `options` is a signature change, albeit implicit. A customization that passes `getParentContextRect` would suddenly break with this change.

Comment (adsa-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2434]: Agreed, I did it just for documenting the `options` parameter (As you told me in the previous conversation). Should we leave documenting the parameters and only change in the docstring or am I missing something ?

Comment (adsa-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js]: I've done the changes, can you please have a look ?

**Comment by Zynton** [comment]
@adsa-odoo I'm a little confused by the disappearance of the line I asked you about just before. :thinking: Also please remove the capital "H" in the title of the second commit.
Other than that, given the amount of changes made post-review, I think this should go through testing again. If it passes then we're good, that's a nice and clean fix.

**Comment by adsa-odoo** [comment]
@Zynton Let me explain about the line which I removed from `_resetPosition` :smile: .

Right now, we are passing `this.options` which is an object containing `getContextFromParentRect` method along with other properties and methods (but we are intrested in one method only).
The change made by me would be required only if I would have used different key. But we are not gonna do that now. So right now no changes are required here. 
Alternatively what we can do is we can replace it by passing `{ getContextFromParentRect: this.options.getContextFromParentRect }` as third argument since we only care about one method and not the other options. (But that's an optional change indeed) .

**Comment by Zynton** [comment]
That would have been a good change but indeed not for the context of the fix. I was only confused because I thought your last comment was saying the change was necessary to fix a bug. Since that's not the case finally, great! If this passes testing I think we'll be good to go :+1:

**Comment by Zynton** [comment]
@robodoo rebase-ff r+

**Comment by adsa-odoo** [comment]
@Zynton Hello, we have to keep in mind that in 16.0 and above we don't have `this.commandbarTablePicker`, we have `this.powerboxTablePicker` otherwise it will through traceback!

**Comment by Zynton** [comment]
@adsa-odoo Thanks for the heads up! Can you make sure to adapt it in the forward port?
